exercises/sgf-parsing/canonical-data.json

```diff
@@ -170,6 +170,171 @@
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "a780b97e-8dbb-474e-8f7e-4031902190e8",
+            "description": "Semicolon in property value doesn't need to be escaped",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[a;b]B[foo];C[bar])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["a;b"],
+                    "B": ["foo"]
+                },
+                "children": [
+                    {
+                        "properties": {
+                            "C": ["bar"]
+                        },
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "0b57a79e-8d89-49e5-82b6-2eaaa6b88ed7",
+            "description": "Parentheses in property value don't need to be escaped",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[x(y)z]B[foo];C[bar])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["x(y)z"],
+                    "B": ["foo"]
+                },
+                "children": [
+                    {
+                        "properties": {
+                            "C": ["bar"]
+                        },
+                        "children": []
+                    }
+                ]
+            }
+        },
+        {
+            "uuid": "3a1023d2-7484-4498-8d73-3666bb386e81",
+            "description": "Escaped newline in property value is converted to nothing at all",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[hello\\\nworld])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["helloworld"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "25abf1a4-5205-46f1-8c72-53273b94d009",
+            "description": "Escaped t and n in property value are just letters, not whitespace",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[\\t = t and \\n = n])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["t = t and n = n"]
+                },
+                "children": []
+            }
+        },
+        {
+            "uuid": "08e4b8ba-bb07-4431-a3d9-b1f4cdea6dab",
+            "reimplements": "11c36323-93fc-495d-bb23-c88ee5844b8c",
+            "description": "mixing various kinds of whitespace and escaped characters in property value",
+            "property": "parse",
+            "input": {
+                "encoded": "(;A[\\]b\nc\\\nd\t\te\\\\ \\\n\\]])"
+            },
+            "expected": {
+                "properties": {
+                    "A": ["]b\ncd  e\\ ]"]
```

Comment:
To be very clear: In this case, and in all other cases in this PR as well, it's important to keep in mind the difference between the string as it appears in JSON and the string as it is presented to the SGF parser.

The string `"(;A[\\]b\nc\\\nd\t\te\\\\ \\\n\\]])"` in JSON means that this string is presented to the SGF parser:

```text
(;A[\]b
c\
d		e\\ \
\]])
```

Everything between the `[` and the `]` is the property value, and it contains these characters:

* `\]`, escaped closing bracket: insert closing bracket in property value
* `b`: insert as-is.
* unescaped newline: insert as-is.
* `c`: insert as-is.
* escaped newline: insert nothing.
* `d`: insert as-is
* two tabs: insert two spaces
* `e`: insert as-is
* `\\`, escaped backslash: insert backslash
* space: insert as-is.
* escaped newline: insert nothing
* `\]`, escaped closing bracket: insert closing bracket

The property value is thus:

```text
]b
cd  e\ ]
```

which is written in JSON as it is on line 333.

Reply:
Would it perhaps be worth it to add this comment to the test case? It looks to be very useful in getting people to understand what they're working with.

Reply:
Not a bad idea. Added.